File: tasks/builtin/bug-fix-04-go/starter.go
package solution

import (
	"fmt"
)

type QueryBuilder struct {
	table          string
	allowedColumns []string
	columns        []string
	whereClause    string
	whereParams    []interface{}
	orderBy        string
}

func NewQueryBuilder(table string, allowedColumns []string) *QueryBuilder {
	return &QueryBuilder{
		table:          table,
		allowedColumns: allowedColumns,
		columns:        []string{"*"},
		whereParams:    []interface{}{},
	}
}

func (qb *QueryBuilder) Select(columns ...string) *QueryBuilder {
	if len(columns) > 0 {
		qb.columns = columns
	} else {
		qb.columns = []string{"*"}
	}
	return qb
}

func (qb *QueryBuilder) Where(condition string, value interface{}) *QueryBuilder {
	if value != nil {
		qb.whereClause = fmt.Sprintf("%s = '%v'", condition, value)
		qb.whereParams = []interface{}{}
	} else {
		qb.whereClause = condition
	}
	return qb
}

func (qb *QueryBuilder) OrderBy(column string, direction string) *QueryBuilder {
	if direction == "" {
		direction = "ASC"
	}
	qb.orderBy = fmt.Sprintf("%s %s", column, direction)
	return qb
}

func (qb *QueryBuilder) Build() (string, []interface{}, error) {
	cols := ""
	for i, col := range qb.columns {
		if i > 0 {
			cols += ", "
		}
		cols += col
	}

	sql := fmt.Sprintf("SELECT %s FROM %s", cols, qb.table)

	params := []interface{}{}
	if qb.whereClause != "" {
		sql += fmt.Sprintf(" WHERE %s", qb.whereClause)
		params = qb.whereParams
	}

	if qb.orderBy != "" {
		sql += fmt.Sprintf(" ORDER BY %s", qb.orderBy)
	}

	return sql, params, nil
}

File: tasks/builtin/bug-fix-04-cs/Starter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class QueryBuilder
{
    private readonly string _table;
    private readonly List<string>? _allowedColumns;
    private List<string> _columns = new() { "*" };
    private string? _whereClause;
    private List<object> _whereParams = new();
    private string? _orderBy;

    public QueryBuilder(string table, List<string>? allowedColumns = null)
    {
        _table = table;
        _allowedColumns = allowedColumns;
    }

    public QueryBuilder Select(params string[] columns)
    {
        _columns = columns.Length > 0 ? columns.ToList() : new List<string> { "*" };
        return this;
    }

    public QueryBuilder Where(string condition, object? value = null)
    {
        if (value != null)
        {
            _whereClause = $"{condition} = '{value}'";
            _whereParams = new List<object>();
        }
        else
        {
            _whereClause = condition;
        }
        return this;
    }

    public QueryBuilder OrderBy(string column, string direction = "ASC")
    {
        _orderBy = $"{column} {direction}";
        return this;
    }

    public (string Sql, List<object> Parameters) Build()
    {
        var cols = string.Join(", ", _columns);
        var sql = $"SELECT {cols} FROM {_table}";

        var parameters = new List<object>();
        if (_whereClause != null)
        {
            sql += $" WHERE {_whereClause}";
            parameters = _whereParams;
        }

        if (_orderBy != null)
        {
            sql += $" ORDER BY {_orderBy}";
        }

        return (sql, parameters);
    }
}
